src/cbmc/bmc_util.cpp

```diff
@@ -0,0 +1,267 @@
+  switch(ui_message_handler.get_ui())
+  {
+  case ui_message_handlert::uit::PLAIN:
+    msg.result() << "Counterexample:" << messaget::eom;
+    show_goto_trace(msg.result(), ns, goto_trace, trace_options);
+    msg.result() << messaget::eom;
+    break;
+
+  case ui_message_handlert::uit::XML_UI:
+  {
+    xmlt xml;
+    convert(ns, goto_trace, xml);
+    msg.status() << xml;
+  }
+  break;
+
+  case ui_message_handlert::uit::JSON_UI:
+  {
+    json_stream_objectt &json_result =
+      ui_message_handler.get_json_stream().push_back_stream_object();
+    const goto_trace_stept &step = goto_trace.steps.back();
+    json_result["property"] =
+      json_stringt(step.pc->source_location.get_property_id());
+    json_result["description"] =
+      json_stringt(step.pc->source_location.get_comment());
+    json_result["status"] = json_stringt("failed");
+    json_stream_arrayt &json_trace =
+      json_result.push_back_stream_array("trace");
+    convert<json_stream_arrayt>(ns, goto_trace, json_trace, trace_options);
+  }
+  break;
+  }
+}
+
+/// outputs witnesses in graphml format
+void output_graphml(
+  safety_checkert::resultt result,
+  const goto_tracet &goto_trace,
+  const symex_target_equationt &symex_target_equation,
+  const namespacet &ns,
+  const optionst &options)
+{
+  const std::string graphml = options.get_option("graphml-witness");
+  if(graphml.empty())
+    return;
+
+  graphml_witnesst graphml_witness(ns);
+  if(result == safety_checkert::resultt::UNSAFE)
+    graphml_witness(goto_trace);
+  else if(result == safety_checkert::resultt::SAFE)
+    graphml_witness(symex_target_equation);
+  else
+    return;
+
+  if(graphml == "-")
+    write_graphml(graphml_witness.graph(), std::cout);
+  else
+  {
+    std::ofstream out(graphml);
+    write_graphml(graphml_witness.graph(), out);
+  }
+}
+
+void convert_symex_target_equation(
+  symex_target_equationt &equation,
+  prop_convt &prop_conv,
+  message_handlert &message_handler)
+{
+  messaget msg(message_handler);
+  msg.status() << "converting SSA" << messaget::eom;
+
+  // convert SSA
+  equation.convert(prop_conv);
+}
+
+std::unique_ptr<memory_model_baset>
+get_memory_model(const optionst &options, const namespacet &ns)
+{
+  const std::string mm = options.get_option("mm");
+
+  if(mm.empty() || mm == "sc")
+    return util_make_unique<memory_model_sct>(ns);
+  else if(mm == "tso")
+    return util_make_unique<memory_model_tsot>(ns);
+  else if(mm == "pso")
+    return util_make_unique<memory_model_psot>(ns);
+  else
+  {
+    throw "invalid memory model '" + mm + "': use one of sc, tso, pso";
+  }
+}
+
+void setup_symex(
+  symex_bmct &symex,
+  const namespacet &ns,
+  const optionst &options,
+  ui_message_handlert &ui_message_handler)
+{
+  messaget msg(ui_message_handler);
+  const symbolt *init_symbol;
+  if(!ns.lookup(INITIALIZE_FUNCTION, init_symbol))
+    symex.language_mode = init_symbol->mode;
+
+  msg.status() << "Starting Bounded Model Checking" << messaget::eom;
```

Comment:
Is this message even true? Symex isn't really BMC any more than any of the prior passes were